### PR TITLE
Week six example code

### PR DESCRIPTION
**Pull request by JasonGibson274**
@chsahit I made the findBlue.cpp code compile and it seems to work, double check me to make sure I did not break anything without realizing it. I have not checked this code working on virtualbox yet. I will update the label when I do.

### CODE REVIEW

File: code/week6/findBlue.cpp
Comment (jgkamat) [inline code/week6/findBlue.cpp]: Right now, closing the window does not quit the program, is it possible to do that? It's a little annoying to close the way it's set up currently.

Comment (JasonGibson274) [reply code/week6/findBlue.cpp]: now it will close if any key is pressed but closing the window will do nothing. Is that acceptable?

Comment (jgkamat) [reply code/week6/findBlue.cpp]: yup, that sounds good to me, at least. :smile:

**Comment by JasonGibson274** [comment]
works on VM provided that extensions pack is installed.

File: code/week6/main.cpp
Comment (jgkamat) [inline code/week6/main.cpp:1]: I think this is the first time we're using Cmake in the course, can you add a list of commands to run to this file (or add a makefile?)

```
mkdir build
cd build
cmake ../
make -j
./findBlue
```

File: code/week6/findBlue.cpp
Comment (jgkamat) [inline code/week6/findBlue.cpp]: For me this causes the image to never come up at all. You probably want:

`waitKey(1) == -1`.

I also take back this being an OK solution, since you might want to press a key and have the program keep running (for example, my window dragging workflow needs to have a key pressed, so I'm unable to manipulate the window).

Perhaps, you could use:
```
(char)waitKey(1) != 'q'
```
or similar.

**Review by jgkamat — CHANGES_REQUESTED** [review]
Just a few more usability fixes

File: code/week6/main.cpp
Comment (barulicm) [reply code/week6/main.cpp:1]: No, we'll be using CMake for all of the hardware application demos for every week. They'll be familiar with this workflow by this point in the semester.

File: code/week6/findBlue.cpp
Comment (barulicm) [reply code/week6/findBlue.cpp]: While the suggested solution is a better way to implement this functionality in general, boo to non-standard window managers. Specifically, don't worry too much about them, since none of the students will be using them.

Comment (jgkamat) [reply code/week6/findBlue.cpp]: >Specifically, don't worry too much about them, since none of the students will be using them. 

There was one student last year in robocup training (who left unfortunately) who used i3 (and many returning robocup team members, including will, at least partially use i3 or other tiling window managers).

There was an incoming robocup member who used emacs + unity to make his own mouseless workflow (which was what inspired me to make my own).

> boo to non-standard window managers

Assuming you're talking about unity (since it seems to be dominant at rj) unity is [officially deprecated by ubuntu](https://insights.ubuntu.com/2017/04/05/growing-ubuntu-for-cloud-and-iot-rather-than-phone-and-convergence/) and will be dropped in future ubuntu versions.

While we shouldn't prioritize alternative workflows, we should solve breaking issues as they are discovered, to avoid alienating future members. I would be happy to do this if no one else is willing to, of course. Diversity in all it's forms is important for organizations to have a balanced viewpoint.

Comment (JasonGibson274) [reply code/week6/findBlue.cpp]: so interestingly enough my computer returns 255 when nothing is pressed.

Comment (jgkamat) [reply code/week6/findBlue.cpp]: Interesting. It's probably something with an implcit cast to unsigned car not happning on my machine.

**Review by jgkamat — DISMISSED** [review]
LGTM, but I don't know too much about opencv, so you probably want another opinion.

I tested this on opencv3 and 2, and it seems to work on both.

**Comment by ebarnette** [comment]
@barulicm can you give this a review before it's merged?

**Comment by JasonGibson274** [comment]
just missing roboracing video @chsahit

**Comment by JasonGibson274** [comment]
@barulicm Does this look better? I slowed the video down a lot so that you can actually see what was happening by comparing the two images but I can speed it up if that is preferred.

File: code/week6/findBlue.cpp
Comment (barulicm) [inline code/week6/findBlue.cpp]: *encoded (not encodeS)

File: code/week6/findBlue.cpp
Comment (barulicm) [inline code/week6/findBlue.cpp]: each pixel is only one *byte*

File: code/week6/findBlue.cpp
Comment (barulicm) [inline code/week6/findBlue.cpp]: No real need to mention ROS in this demo

File: code/week6/findBlue.cpp
Comment (barulicm) [inline code/week6/findBlue.cpp]: This loads from a video, not the default video device.

File: code/week6/findBlue.cpp
Comment (barulicm) [inline code/week6/findBlue.cpp]: Do we want to use `std::cerr`?

File: code/week6/findBlue.cpp
Comment (barulicm) [inline code/week6/findBlue.cpp]: In any real-world code this would be declared outside of the loop for performance reasons

File: code/week6/main.cpp
Comment (barulicm) [inline code/week6/main.cpp]: Why is this inside the main method? Also we should be consistent about which namespaces we include. The findBlue demo uses cv and makes std explicit, this demo does the opposite.
I recommend not including the cv namespace so it is obvious which functions are from OpenCV as opposed to standard c++.

File: code/week6/main.cpp
Comment (barulicm) [inline code/week6/main.cpp]: This is not the fastest way to iterate over an image in OpenCV. Do we care?

File: code/week6/main.cpp
Comment (barulicm) [inline code/week6/main.cpp]: *main()* reaches the end without returning anything.

**Review by barulicm — CHANGES_REQUESTED** [review]

File: code/week6/main.cpp
Comment (JasonGibson274) [reply code/week6/main.cpp]: more for my own benefit here. Isn't return 0 implicitly added anyway so why is it worth making it explicit?

File: code/week6/main.cpp
Comment (JasonGibson274) [reply code/week6/main.cpp]: do we want an example of iterating over a Mat or should this just be taken out?

File: code/week6/findBlue.cpp
Comment (JasonGibson274) [reply code/week6/findBlue.cpp]: Can you explain the performance issue here?

File: code/week6/main.cpp
Comment (barulicm) [reply code/week6/main.cpp]: That's up to someone else. It's a useful thing to know, but you can also do a lot without using it.

**Comment by JasonGibson274** [comment]
@barulicm final checks

File: code/week6/findBlue.cpp
Comment (barulicm) [inline code/week6/findBlue.cpp]: Why do we need unisdt.h here?

File: code/week6/main.cpp
Comment (barulicm) [inline code/week6/main.cpp]: Use current standard headers ie <string>

File: code/week6/main.cpp
Comment (barulicm) [inline code/week6/main.cpp]: Why do we need stdlib.h here?

File: code/week6/main.cpp
Comment (barulicm) [reply code/week6/main.cpp]: This comment still stands

File: code/week6/main.cpp
Comment (barulicm) [inline code/week6/main.cpp]: Noticing a lack of cv:: here, does this compile?

**Review by barulicm — CHANGES_REQUESTED** [review]
Some of my comments from my last review were not addressed.
Be careful with what appear to be includes suggested by your IDE. C++ standard headers like <string> and <iostream> are preferred over C or system-specific headers.

File: code/week6/main.cpp
Comment (JasonGibson274) [reply code/week6/main.cpp]: This actually does compile on my machine but I will add it explicitly.

File: code/week6/main.cpp
Comment (JasonGibson274) [reply code/week6/main.cpp]: I added something along the lines of this http://docs.opencv.org/2.4/doc/tutorials/core/how_to_scan_images/how_to_scan_images.html#howtoscanimagesopencv

**Comment by JasonGibson274** [comment]
@barulicm sorry I missed those changes in mentioned earlier. If there is a faster/better way to iterate over an image let me know that is what I found on the opencv docs.

**Review by barulicm — APPROVED** [review]

**Pull request merged**
